File: cpp/src/mnistloader.cpp
#include "mnistloader.h"
#include <arpa/inet.h>
#include <boost/iostreams/filter/gzip.hpp>
#include <boost/iostreams/filtering_streambuf.hpp>
#include <fstream>
#include <iostream>
#include <stdint.h>
#include <string>
#include <vector>

using namespace std;
using namespace boost::iostreams;


//
// Read MNIST digit labels from the given file at path.
//
LABELS
readLabelFile(string const &path)
{
  ifstream file(path, ios::binary);
  filtering_streambuf<input> in;
  in.push(gzip_decompressor());
  in.push(file);
  istream stream(&in);
  stream.exceptions(ios::failbit | ios::badbit);

  uint32_t magicNum;
  stream.read(reinterpret_cast<char*>(&magicNum), sizeof(magicNum));
  magicNum = ntohl(magicNum);
  if (magicNum != 2049) {
    throw runtime_error(path + " has wrong magic number");
  }
  uint32_t nLabels;
  stream.read(reinterpret_cast<char*>(&nLabels), sizeof(nLabels));
  nLabels= ntohl(nLabels);

  LABELS labels;
  uint8_t label;
  for (uint32_t i=0; i < nLabels; i++) {
    stream.read(reinterpret_cast<char*>(&label), sizeof(label));
    labels.emplace_back(label);
  }
  return labels;
}


//
// Read MNIST digit images from the file at path.
//
IMAGES
readImageFile(string const &path)
{
  ifstream file(path, ios::binary);
  filtering_streambuf<input> in;
  in.push(gzip_decompressor());
  in.push(file);
  istream stream(&in);
  stream.exceptions(ios::failbit | ios::badbit);

  uint32_t magicNum;
  stream.read(reinterpret_cast<char*>(&magicNum), sizeof(magicNum));
  magicNum = ntohl(magicNum);
  if (magicNum != 2051) {
    throw runtime_error(path + " has wrong magic number");
  }
  uint32_t nImages;
  stream.read(reinterpret_cast<char*>(&nImages), sizeof(nImages));
  nImages = ntohl(nImages);
  uint32_t nRows;
  stream.read(reinterpret_cast<char*>(&nRows), sizeof(nRows));
  nRows = ntohl(nRows);
  uint32_t nCols;
  stream.read(reinterpret_cast<char*>(&nCols), sizeof(nCols));
  nCols = ntohl(nCols);

  IMAGES images;
  uint8_t pixel;
  for (uint32_t i=0; i < nImages; i++) {
    IMG image;
    for (uint32_t j=0; j < (nRows*nCols); j++) {
        stream.read(reinterpret_cast<char*>(&pixel), sizeof(pixel));
        image.emplace_back(pixel);
    }
    images.emplace_back(image);
  }
  return images;
}


//
// Read MNIST trainings data from the given directory at path.
//
pair<IMAGES, LABELS>
loadMnistTrainingData(string const &path)
{
  auto img = readImageFile(path + "/train-images-idx3-ubyte.gz");
  auto lbl = readLabelFile(path + "/train-labels-idx1-ubyte.gz");
  return make_pair(img, lbl);
}


//
// Read MNIST validation data from the given directory at path.
//
pair<IMAGES, LABELS>
loadMnistValidationData(string const &path)
{
  auto img = readImageFile(path + "/t10k-images-idx3-ubyte.gz");
  auto lbl = readLabelFile(path + "/t10k-labels-idx1-ubyte.gz");
  return make_pair(img, lbl);
}

File: haskell/src/MnistLoader.hs
{-# LANGUAGE BangPatterns #-}
module MnistLoader
  ( load_data_wrapper
  , TrainingData
  , ValidationData
  ) where

import           Codec.Compression.GZip (decompress)
import           Control.Applicative ((<$>))
import           Control.DeepSeq
import           Control.Monad (replicateM, unless)
import           Data.Binary.Get
import qualified Data.ByteString.Lazy as B
import           Data.Word (Word8)
import           Numeric.LinearAlgebra
import           System.FilePath.Posix ((</>))

type TrainingData   = [(Vector Double, Vector Double)]
type ValidationData = [(Vector Double, Word8)]


readImageFile :: FilePath -> IO [[Word8]]
readImageFile p = runGet parseImages <$> decompress <$> B.readFile p


readLabelFile :: FilePath -> IO [Word8]
readLabelFile p = runGet parseLabels <$> decompress <$> B.readFile p


parseImages :: Get [[Word8]]
parseImages = do
  magicNum <- getWord32be
  unless (magicNum == 2051) $ error $
    "Not an image file, wrong magic number " ++ show magicNum
  nImg <- fromIntegral <$> getWord32be
  nRow <- fromIntegral <$> getWord32be
  nCol <- fromIntegral <$> getWord32be
  img  <- replicateM nImg $ replicateM (nRow * nCol) getWord8
  unlessM isEmpty $ error "not all data consumed"
  return img


parseLabels :: Get [Word8]
parseLabels = do
  magicNum <- getWord32be
  unless (magicNum == 2049) $ error $
    "Not a label file, wrong magic number " ++ show magicNum
  n <- fromIntegral <$> getWord32be
  lbl <- replicateM n getWord8
  unlessM isEmpty $ error "not all data consumed"
  return lbl


load_data_wrapper :: FilePath -> IO (TrainingData, ValidationData)
load_data_wrapper p = do
  imgTraining    <- readImageFile $ p </> "train-images-idx3-ubyte.gz"
  imgValidation  <- readImageFile $ p </> "t10k-images-idx3-ubyte.gz"
  lblTraining    <- readLabelFile $ p </> "train-labels-idx1-ubyte.gz"
  lblValidation  <- readLabelFile $ p </> "t10k-labels-idx1-ubyte.gz"
  return ( (zip $!! normalizeImg imgTraining)   $!! toUnitVector lblTraining
         , (zip $!! normalizeImg imgValidation) $!! lblValidation
         )
  where
    normalizeImg = map $ fromList . map ((/ 256) . fromIntegral)
    toUnitVector = map (fromList . toUnitVector')
    toUnitVector' 0 = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    toUnitVector' 1 = [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    toUnitVector' 2 = [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]
    toUnitVector' 3 = [0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
    toUnitVector' 4 = [0, 0, 0, 0, 1, 0, 0, 0, 0, 0]
    toUnitVector' 5 = [0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
    toUnitVector' 6 = [0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
    toUnitVector' 7 = [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
    toUnitVector' 8 = [0, 0, 0, 0, 0, 0, 0, 0, 1, 0]
    toUnitVector' 9 = [0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
    toUnitVector' x = error $ "(toUnitVector " ++ show x ++ ") is undefined"


unlessM :: Monad m => m Bool -> m () -> m ()
unlessM mp f = (\p -> unless p f) =<< mp
